virtualmachine/vsphere/util.go

```diff
@@ -875,6 +875,7 @@ var cloneFromTemplate = func(vm *VM, dcMo *mo.Datacenter, usableDatastores []str
 		config.DeviceChange = append(config.DeviceChange, conf...)
 	}
 
+	checkCustomSpecMutex.Lock()
 	err = checkAndCreateCustomSpec(vm)
 	if err != nil {
 		return fmt.Errorf("Error creating custom spec: %v", err)
```

Comment:
In case of error, function will return without unlocking the mutex, and we will have a deadlock. We should unlock and then return.
Applicable on line 889 also.